src/lib/utils.ts

```diff
@@ -44,3 +24,84 @@ export function encodeBase32(buf: Buffer): string {
   const encoder = new base32.Encoder({ type: 'crockford', lc: true })
   return encoder.write(buf).finalize()
 }
+
+/**
+ * Generate an idempotent unique id for a given tilejson. Not all tilejson has
+ * an id field, so we use the tile URL as an identifier (assumes two tilejsons
+ * refering to the same tile URL are the same)
+ */
+export function getTilesetId(tilejson: TileJSON): string {
+  // If the tilejson has no id, use the tile URL as the id
+  const id = tilejson.id || tilejson.tiles.sort()[0]
+  return encodeBase32(hash(id))
+}
+
+/**
+ * Get the upstream tile URL for a particular tile
+ */
+export function getInterpolatedUpstreamTileUrl({
+  tiles: templateUrls,
+  scheme: upstreamScheme = 'xyz',
+  zoom,
+  x,
+  y,
+}: {
+  tiles: TileJSON['tiles']
+  scheme: TileJSON['scheme']
+  zoom: number
+  x: number
+  y: number
+}): string | undefined {
+  // TODO: Support {ratio} in template URLs, not used in mapbox-gl-js, only in
+  // the mobile SDKs
+  const ratio = ''
+
+  if (!isStringArray(templateUrls)) {
+    console.log('templateUrls', templateUrls)
+    return
+  }
+
+  const bbox = getTileBBox(x, y, zoom)
+  const quadkey = tileToQuadKey({ x, y, zoom })
+
+  return templateUrls[(x + y) % templateUrls.length]
+    .replace('{prefix}', (x % 16).toString(16) + (y % 16).toString(16))
+    .replace('{z}', String(zoom))
+    .replace('{x}', String(x))
+    .replace(
+      '{y}',
+      String(upstreamScheme === 'tms' ? Math.pow(2, zoom) - y - 1 : y)
```

Comment:
note to self from Gregor: we should probably convert this earlier and use the converted y value for the `bbox` and `quadkey` calculations too. in general, need to think about how the scheme affects other calculations

Reply:
hm actually looks like `getTileBBox` is accounting for that already, so no need to convert `y` before calling that https://github.com/mapbox/whoots-js/blob/63b423ee084d47713256f9b0e310a0d3bbeeba64/index.mjs#L55

Reply:
The calculation will be different depending on the scheme. We need to know whether these functions expect tms or XYZ.

Reply:
ah yes i just realized that the calculation is the same for going both ways, so yeah will dig in further